[DOC-10638] Add LATERAL keyword support

julia-cb opened the pull request:
Adding LATERAL keyword support to ANSI JOIN, ANSI NEST, and Comma-Separated Join.

Jira ticket: https://issues.couchbase.com/browse/DOC-10638

Preview doc pages:

- [JOIN Clause](https://preview.docs-test.couchbase.com/lateralkeyword2/server/current/n1ql/n1ql-language-reference/join.html)
- [NEST Clause](https://preview.docs-test.couchbase.com/lateralkeyword2/server/current/n1ql/n1ql-language-reference/nest.html)
- [Comma-Separated Join](https://preview.docs-test.couchbase.com/lateralkeyword2/server/current/n1ql/n1ql-language-reference/comma.html)

----///----

Changes in this PR:

- [ANSI JOIN](https://docs.couchbase.com/server/current/n1ql/n1ql-language-reference/join.html#section_ek1_jnx_1db)
   - Added new section on LATERAL Joins
   - Added new LATERAL example
   - Updated ansi-join-clause ebnf file and diagram to include LATERAL
   - Created ansi-join-lateral n1ql and jsonc files for LATERAL example
 
- [ANSI NEST](https://docs.couchbase.com/server/current/n1ql/n1ql-language-reference/nest.html#section_tc1_nnx_1db)
   - Added new section on LATERAL Nest
   - Added new LATERAL example
   - Updated ansi-nest-clause ebnf file and diagram to include LATERAL
   - Created ansi-nest-lateral n1ql and jsonc files for LATERAL example

- [Comma-Separated Join](https://docs.couchbase.com/server/current/n1ql/n1ql-language-reference/comma.html)
   - Added new section on LATERAL Join
   - Added LATERAL query to existing "Cartesian product" example (Example 1)
   - Updated comma-separated-join ebnf file and diagram to include LATERAL

julia-cb committed:
Added LATERAL to example

julia-cb committed:
Added LATERAL to example

julia-cb committed:
Added LATERAL to example

julia-cb committed:
Added LATERAL to example

julia-cb committed:
Added LATERAL to ANSI JOIN Clause section of page

simon-dew commented on modules/n1ql/partials/grammar/dql.ebnf:
```diff
@@ -133,7 +133,7 @@ join-clause ::= ansi-join-clause | lookup-join-clause | index-join-clause
 
 /* tag::ansi-join-clause[] */
-ansi-join-clause ::= ansi-join-type? 'JOIN' ansi-join-rhs ansi-join-predicate
+ansi-join-clause ::= ansi-join-type? 'JOIN LATERAL' ansi-join-rhs ansi-join-predicate
```
Good start! 'JOIN' is mandatory in the JOIN clause, but 'LATERAL' is optional, so these two tokens need to be separate, and the LATERAL keyword needs to be marked as optional with a question mark:

```suggestion
ansi-join-clause ::= ansi-join-type? 'JOIN' 'LATERAL'? ansi-join-rhs ansi-join-predicate
```

simon-dew commented on modules/n1ql/examples/select/ansi-join-cover.n1ql:
```diff
@@ -10,7 +10,7 @@ ON route (sourceairport, destinationairport, airline);
 -- tag::query[]
 SELECT META(route).id route_id, route.airline, route.destinationairport
-FROM airport JOIN route ON route.sourceairport = airport.faa
+FROM airport JOIN LATERAL route ON route.sourceairport = airport.faa
```
We shouldn't add the LATERAL keyword to the existing examples, because it strongly implies that LATERAL is mandatory, when it is (a) optional and (b) doesn't do anything.

Instead we should add a single new example, which is exactly the same as one of the existing examples (*), but with the inclusion of the LATERAL keyword. The new example would point out that the results of the new example are exactly the same.

(*) I would suggest `ansi-join-inner.n1ql`

simon-dew commented on modules/n1ql/examples/select/ansi-join-inner.n1ql:
```diff
@@ -1,6 +1,6 @@
 SELECT route.airlineid, airline.name, route.sourceairport, route.destinationairport
 FROM route
-INNER JOIN airline
+INNER JOIN LATERAL airline
```
Revert (as per previous comment)

simon-dew commented on modules/n1ql/examples/select/ansi-join-left.n1ql:
```diff
@@ -2,9 +2,9 @@ SELECT DISTINCT  MIN(aport.airportname) AS Airport__Name,
                  MIN(lmark.name) AS Landmark_Name
 FROM airport aport -- <.>
-LEFT JOIN landmark lmark -- <.>
+LEFT JOIN LATERAL landmark lmark -- <.>
```
Revert (as per previous comment)

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/join.adoc:
```diff
@@ -108,7 +108,7 @@ a|
 SELECT *
 FROM route
-JOIN airline
+JOIN LATERAL airline
```
Revert (as per previous comment)

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/join.adoc:
```diff
@@ -118,7 +118,7 @@ a|
 SELECT *
 FROM route
-LEFT JOIN airline
+LEFT JOIN LATERAL airline
```
Revert (as per previous comment)

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/join.adoc:
```diff
@@ -170,6 +170,8 @@ This expression may contain fields, constant expressions, or any complex {sqlpp}
 
 * You may chain ANSI joins with comma-separated joins; however, the comma-separated joins must come after any JOIN, NEST, or UNNEST clauses.
 
+* INNER joins and LEFT OUTER joins support the LATERAL keyword in front of the right-hand side keyspace. RIGHT OUTER joins do not support the LATERAL keyword.
```
Good start on the description, but this is a new keyword, not a limitation, so I don't think it belongs in this section. We aim to describe the keywords (as far as possible) in the order they appear in the syntax summary. So there should be a new heading 4 section called something like `==== LATERAL Joins` between the sections on Join Type and Join Predicate &mdash; i.e. starting at line 148 above.

simon-dew reviewed (changes requested):
Good start! Comments to keep you on the right track

julia-cb committed:
Updated info based on Simon's feedback and added new n1ql and jsonc files for ansi-join-lateral

julia-cb committed:
Updated ansi-join-clause image

julia-cb commented:
@simon-dew Thanks for the feedback both here and over zoom, Simon! I updated the JOIN page with your feedback (and was even able to update the ansi-join-clause image!). I'll update the NEST and Comma pages with the LATERAL info and add it to the this PR :)

I also moved this PR back into draft mode.

julia-cb committed:
Removed LATERAL from ansi-join-cover n1ql file

julia-cb committed:
Added LATERAL info to NEST page and created new ansi-nest-lateral n1ql and jsonc files

julia-cb committed:
Removed extra space from page

julia-cb committed:
Added LATERAL information to Comma page and updated ebnf file

julia-cb committed:
Added comma-separated-join updated image

julia-cb commented:
@simon-dew Hi Simon, this is ready for a review whenever you have a moment! Thank you :)

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/join.adoc:
```diff
@@ -145,6 +146,16 @@ In Couchbase Server 6.5 and later, if you create either of the following:
+==== LATERAL Join
+
+Couchbase supports lateral queries with and without the LATERAL keyword.
+This makes Couchbase queries compatible with queries from relational databases.
```
May seem paradoxical, but I think this needs more information on what this keyword is for and why you would want use it. Even though it doesn't do anything. Something like:

```suggestion
When an expression on the right-hand side of an ANSI join references a keyspace that is already specified in the same FROM clause, the expression is said to be correlated.
In relational databases, a join which contains correlated expressions is referred to as a lateral join.
In {sqlpp}, lateral correlations are detected automatically, and there is no need to specify that a join is lateral.

In Couchbase Server 7.6 and later, you can use the LATERAL keyword as a visual reminder that a join contains correlated expressions.
The LATERAL keyword is not required -- the keyword is included solely for compatibility with queries from relational databases.

If you use the LATERAL keyword in a join that has no lateral correlation, the keyword is ignored.
```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/join.adoc:
```diff
@@ -62,6 +62,7 @@ image::n1ql-language-reference/ansi-join-clause.png["Syntax diagram", align=left
 
 [horizontal.compact]
 ansi-join-type:: <<ansi-join-type>> icon:caret-down[]
+ansi-join-lateral:: <<ansi-join-lateral>> icon:caret-down[]
```
Delete this line: there's no element in the syntax diagram called `ansi-join-lateral`

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/join.adoc:
```diff
@@ -215,8 +245,8 @@ include::example$select/ansi-join-left.jsonc[]
 
-[#ANSI-Join-Example-3]
-.RIGHT OUTER JOIN of <<ANSI-Join-Example-2>>
+[#ANSI-Join-Example-4]
```
Revert this. As a general rule I don't change IDs within a topic: you never know what might be linking to them.

```suggestion
[#ANSI-Join-Example-3]
```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/join.adoc:
```diff
@@ -215,8 +245,8 @@ include::example$select/ansi-join-left.jsonc[]
-[#ANSI-Join-Example-3]
-.RIGHT OUTER JOIN of <<ANSI-Join-Example-2>>
+[#ANSI-Join-Example-4]
+.RIGHT OUTER JOIN of <<ANSI-Join-Example-3>>
```
Revert this, as above.

```suggestion
.RIGHT OUTER JOIN of <<ANSI-Join-Example-2>>
```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/join.adoc:
```diff
@@ -237,7 +267,7 @@ include::example$select/ansi-join-right.jsonc[]
 ====
 
-[#ANSI-Join-Example-4]
+[#ANSI-Join-Example-5]
```
Revert this, as above.

```suggestion
[#ANSI-Join-Example-4]
```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/join.adoc:
```diff
@@ -194,6 +205,25 @@ The INNER JOIN only returns results where a left-side document matches a right-s
 ====
 
 [#ANSI-Join-Example-2]
```
Give the new example a new ID

```suggestion
[#ANSI-Join-Example-1A]
```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/join.adoc:
```diff
@@ -194,6 +205,25 @@ The INNER JOIN only returns results where a left-side document matches a right-s
+The INNER LATERAL JOIN returns the same results as <<ANSI-Join-Example-1,Example 1>>.
+====
+
+[#ANSI-Join-Example-3]
```
Revert this, so the existing example keeps the same ID

```suggestion
[#ANSI-Join-Example-2]
```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/comma.adoc:
```diff
@@ -64,6 +64,14 @@ For each joined object produced, both the left-hand side and right-hand side sou
+==== LATERAL Join
+
+Couchbase supports lateral queries with and without the LATERAL keyword.
+This makes Couchbase queries compatible with queries from relational databases.
```
more information on what this keyword is for and why you would want use it

```suggestion
When an expression on the right-hand side of a comma-separated join references a keyspace that is already specified in the same FROM clause, the expression is said to be correlated.
In relational databases, a join which contains correlated expressions is referred to as a lateral join.
In {sqlpp}, lateral correlations are detected automatically, and there is no need to specify that a join is lateral.

In Couchbase Server 7.6 and later, you can use the LATERAL keyword as a visual reminder that a join contains correlated expressions.
The LATERAL keyword is not required -- the keyword is included solely for compatibility with queries from relational databases.

If you use the LATERAL keyword in a join that has no lateral correlation, the keyword is ignored.
```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/nest.adoc:
```diff
@@ -87,6 +88,14 @@ For each nested object produced, only the left-hand source objects must be non-M
+==== LATERAL Nest
+
+Couchbase supports lateral queries with and without the LATERAL keyword.
+This makes Couchbase queries compatible with queries from relational databases.
```
more information on what this keyword is for and why you would want use it

```suggestion
When an expression on the right-hand side of an ANSI nest references a keyspace that is already specified in the same FROM clause, the expression is said to be correlated.
In relational databases, a join which contains correlated expressions is referred to as a lateral join.
In {sqlpp}, lateral correlations are detected automatically, and there is no need to specify that a nest or join is lateral.

In Couchbase Server 7.6 and later, you can use the LATERAL keyword as a visual reminder that a nest contains correlated expressions.
The LATERAL keyword is not required -- the keyword is included solely for compatibility with queries from relational databases.

If you use the LATERAL keyword in a nest that has no lateral correlation, the keyword is ignored.
```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/nest.adoc:
```diff
@@ -167,6 +176,24 @@ include::example$select/ansi-nest-inner.jsonc[tags=extract;ellipsis]
+----
+====
+
+[#ANSI-NEST-Example-3]
```
Revert this -- keep existing IDs the same

```suggestion
[#ANSI-NEST-Example-2]
```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/nest.adoc:
```diff
@@ -167,6 +176,24 @@ include::example$select/ansi-nest-inner.jsonc[tags=extract;ellipsis]
 ====
 
 [#ANSI-NEST-Example-2]
```
Use a new ID for the new example

```suggestion
[#ANSI-NEST-Example-1A]
```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/comma.adoc:
```diff
@@ -191,7 +199,16 @@ SELECT * FROM [{"abc": 1}, {"abc": 2}, {"abc": 3}] AS a,
+              [{"xyz": 1}, {"xyz": 2}] AS b;
+----
+
+Compare the queries above with the following query using an ANSI join.
```
Based on the importance given to comma-separated joins with subqueries in the PRD, I think we need to have a separate example for LATERAL with comma-separated joins. Revert this change, and let's add a separate example at the end of this section.

```suggestion
Compare the query above with the following query using an ANSI join.
```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/comma.adoc:
```diff
@@ -200,7 +217,7 @@ SELECT * FROM [{"abc": 1}, {"abc": 2}, {"abc": 3}] AS a
 ----
 
-The results of the two queries are the same.
+The results of the three queries are the same.
```
Revert this, as above

```suggestion
The results of the two queries are the same.
```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/comma.adoc:
At the end of this page, add a new example as follows:

```asciidoc
[#example-lateral]
.Lateral correlation
====
The following query has a lateral correlation between the subquery and the `airport` keyspace.

.Comma-separated join
[source,sqlpp]
----
SELECT airport.airportname, t2.name
FROM airport,
(SELECT name FROM hotel WHERE hotel.city = airport.city) AS t2
LIMIT 5;
----

Compare the query above with the following query using the LATERAL keyword.

.Comma-separated join with LATERAL keyword
[source,sqlpp]
----
SELECT airport.airportname, t2.name
FROM airport,
LATERAL (SELECT name FROM hotel WHERE hotel.city = airport.city) AS t2
LIMIT 5;
----

The results of the two queries are the same.

.Results
[source,json]
----
[
  {
    "airportname": "Mandelieu",
    "name": "Hotel Cybelle"
  },
  {
    "airportname": "Cote D\\'Azur",
    "name": "Best Western Hotel Riviera Nice"
  },
  {
    "airportname": "Cote D\\'Azur",
    "name": "Hotel Anis"
  },
  {
    "airportname": "Cote D\\'Azur",
    "name": "NH Nice"
  },
  {
    "airportname": "Cote D\\'Azur",
    "name": "Hotel Suisse"
  }
]
----
====
```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/comma.adoc:
```diff
@@ -64,6 +64,14 @@ For each joined object produced, both the left-hand side and right-hand side sou
 The _right-hand side_ of a comma-separated join may be a keyspace reference, a subquery, or a generic expression term.
 
+[#comma-join-lateral]
+==== LATERAL Join
```
In this page, this needs to be a level 3 heading. (This page doesn't have the extra level of ANSI / Lookup / Index that the JOIN and NEST pages do.)

```suggestion
=== LATERAL Join
```

simon-dew commented on modules/n1ql/examples/select/ansi-join-lateral.n1ql:
This example works, but it only illustrates how the LATERAL keyword is ignored if there is no correlation between the left-hand side and the right-hand side of the join. Can you ask @miaobingjie or the SQL++ team if there is an example of an ANSI join with correlation between the left-hand side and the right-hand side of the join, so we can show the LATERAL keyword in the relevant environment?

simon-dew commented on modules/n1ql/examples/select/ansi-join-lateral.jsonc:
May need to update this. See comments on `ansi-join-lateral.n1ql`.

simon-dew commented on modules/n1ql/examples/select/ansi-nest-lateral.n1ql:
This example works, but it only illustrates how the LATERAL keyword is ignored if there is no correlation between the left-hand side and the right-hand side of the nest. Can you ask @miaobingjie or the SQL++ team if there is an example of an ANSI nest with correlation between the left-hand side and the right-hand side of the nest, so we can show the LATERAL keyword in the relevant environment?

simon-dew commented on modules/n1ql/examples/select/ansi-nest-lateral.jsonc:
May need to update this. See comments on `ansi-nest-lateral.n1ql`.

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/join.adoc:
```diff
@@ -145,6 +146,16 @@ In Couchbase Server 6.5 and later, if you create either of the following:
+INNER JOINS and LEFT OUTER JOINS support the optional LATERAL keyword in front of the right-hand side keyspace.
+
+RIGHT OUTER JOINS do not support the optional LATERAL keyword.
+
```
Add a note at the end here:

```asciidoc
NOTE: Using the LATERAL keyword in an ANSI join implies that the right-hand side of the join must appear after the left-hand side of the join.
This may prevent the cost-based optimizer from reordering joins in the query to give the optimal join order.
For details, see xref:n1ql:n1ql-language-reference/cost-based-optimizer.adoc#join-enumeration[Join Enumeration].
```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/comma.adoc:
```diff
@@ -64,6 +64,14 @@ For each joined object produced, both the left-hand side and right-hand side sou
+This makes Couchbase queries compatible with queries from relational databases.
+
+You can use the optional LATERAL keyword in front of the right-hand side keyspace of a comma-separated join.
+
```
Add a note at the end here:

```asciidoc
NOTE: Using the LATERAL keyword in a comma-separated join implies that the right-hand side of the join must appear after the left-hand side of the join.
This may prevent the cost-based optimizer from reordering joins in the query to give the optimal join order.
For details, see xref:n1ql:n1ql-language-reference/cost-based-optimizer.adoc#join-enumeration[Join Enumeration].
```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/comma.adoc:
```diff
@@ -64,6 +64,14 @@ For each joined object produced, both the left-hand side and right-hand side sou
 
+[#comma-join-lateral]
+==== LATERAL Join
+
```
Standard boilerplate text for new feature

```suggestion

ifeval::['{page-component-version}' == '7.6']
_(Introduced in Couchbase Server 7.6)_
endif::[]

```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/join.adoc:
```diff
@@ -145,6 +146,16 @@ In Couchbase Server 6.5 and later, if you create either of the following:
 
+[#ansi-join-lateral]
+==== LATERAL Join
+
```
Standard boilerplate text for new feature

```suggestion

ifeval::['{page-component-version}' == '7.6']
_(Introduced in Couchbase Server 7.6)_
endif::[]

```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/nest.adoc:
```diff
@@ -87,6 +88,14 @@ For each nested object produced, only the left-hand source objects must be non-M
 
+[#ansi-nest-lateral]
+==== LATERAL Nest
+
```
Standard boilerplate text for new feature

```suggestion

ifeval::['{page-component-version}' == '7.6']
_(Introduced in Couchbase Server 7.6)_
endif::[]

```

simon-dew reviewed (changes requested):
Hi @julia-cb ! Thanks for your updates and accurate examples. The only thing was, I don't think they fully explained why a developer would want to use this feature. I went back to the feature PRD to look at the use cases that they were focused on. I've made suggestions in the following areas:

1. Suggested improved wording to explain what the feature is for.
2. Suggested example for comma-separated joins, based on the PRD.
3. Notes about the impact on query optimization, also taken from the PRD.
4. Other minor changes.

Unfortunately this makes it look as if I've completely ripped it to shreds again. I am sorry about that!

simon-dew commented:
When we're finished with peer review, please add @miaobingjie as an SME reviewer (as the person who raised the original ticket)

julia-cb committed:
Apply suggestions from code review

Co-authored-by: Simon Dew <39966290+simon-dew@users.noreply.github.com>

julia-cb committed:
Apply suggestions from code review

Co-authored-by: Simon Dew <39966290+simon-dew@users.noreply.github.com>

julia-cb committed:
Updated with Simon's feedback

julia-cb commented:
@simon-dew Thanks Simon! Great feedback and I applied all of it. Bingjie also gave me updated examples for the n1ql and jsonc files for ANSI JOIN and ANSI NEST.

julia-cb committed:
Update ansi-join-lateral.n1ql with Bingjie's example

julia-cb committed:
Update ansi-join-lateral.jsonc with Bingjie's example

julia-cb committed:
Update ansi-nest-lateral.n1ql with Bingjie's example

julia-cb committed:
Update ansi-nest-lateral.jsonc with Bingjie's example

julia-cb committed:
Added extract tag to file

julia-cb committed:
Added extract tag to file

julia-cb commented:
@simon-dew Hey Simon, let me know if this PR is good to be merged (Bingjie already provided the code examples over slack so I'm not sure a thorough review from him is necessary).

simon-dew reviewed (approved):
LGTM 👍 let's get it merged, and if there are any pernickety changes we can fix them later.
